### ISSUE TITLE
All files show wrong modification date (1.1.1970 / "just now")

### ISSUE THREAD

**Issue by LinusHoppe** [owncloud/core#29287]
### Steps to reproduce
1. Install Owncloud
2. Sync files to Owncloud

### Expected behaviour
The files have the correct modification date and are synced correctly.

### Actual behaviour

- My synced folders show the correct modification date, e. g. "10 days ago" at "08.10.2017 14:50" (see folders.png.
- All files (across all folders) show "just now" as modification date, although I've synced them days ago (see files.png). Having no clients connected at the moment, the date remains incorrect.
- I've connected a second device, the files are downloaded. All files show the modification date "01.01.1970 01:00:00" as modification date (see windows_explorer.png).
- The incorrect date lead to clients wo don't sync any files (changed files are not uploaded to owncloud)

### Images
files.png:
![files](https://user-images.githubusercontent.com/32911160/31737980-85d5368c-b449-11e7-8267-219b775956b6.png)
folders.png:
![folders](https://user-images.githubusercontent.com/32911160/31737986-88ee5484-b449-11e7-8656-5cc62802dc55.png)
windows_explorer.png:
![windows_explorer](https://user-images.githubusercontent.com/32911160/31737990-8a960534-b449-11e7-978d-5684dbdf752a.png)

### Server configuration
**Operating system**:
Linux

**Web server:**
Managed Webspace, Apache httpD

**Database:**
MySQL 

**PHP version:**
5.6.30

**ownCloud version:** (see ownCloud admin page)
ownCloud 10.0.3 (stable) 

**Updated from an older ownCloud or fresh install:**
fresh install

**Where did you install ownCloud from:**
ownCloud Web Installer (https://owncloud.org/install/)

**Signing status (ownCloud 9.0 and above):**
https://gist.github.com/GeneralAmbush/5c0989888a4b8273086d14ab2896622d

**The content of config/config.php:**
https://gist.github.com/GeneralAmbush/af4d5ecb8594d8f5298175c75cb14e62

**Are you using external storage, if yes which one:** local/smb/sftp/...
No

**Are you using encryption:** yes/no
Yes

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
No

### Client configuration
**Browser:**
Firefox, Chrome, IE

**Operating system:**
Windows 8.1

#### ownCloud log (data/owncloud.log)
My ownCloud log is 10 MB because of many lines of 
```
{"reqId":"6syFlxfihjPDI2087ARm","level":3,"time":"2017-10-18T19:02:22+00:00","remoteAddr":"**myIP**","user":"**myUser**","app":"PHP","method":"GET","url":"\/owncloud\/ocs\/v2.php\/apps\/notifications\/api\/v1\/notifications?format=json","message":"You are using a fallback implementation of the intl extension. Installing the native one is highly recommended instead. at /owncloud\/lib\/composer\/patchwork\/utf8\/src\/Patchwork\/Utf8\/Bootup\/intl.php#18"}
```

**Comment by PVince81** [comment]
Pick a file with wrong timestamp and do `select * from oc_filecache where path like '%pickedfile%'` and check the mtime and storage_mtime columns. For some reason, something might have written a 0 there.

Considering that you are using managed web space, I'd suspect some environment related issue there.

**Comment by GeneralAmbush** [comment]
@PVince81: You are right. storage_mtime is always 0. mtime is always 0 if the path is a file. If the path is a folder, mtime seems to be correct.

I'm not sure about the environment related issue. I've been using owncloud on this webspace since a few years, just after the re-install the error occurs. I've talked to my webhoster, they haven't changed anything and there don't know any issues with owncloud on their environment.

**Comment by PVince81** [comment]
Any third party apps enabled ?

Currently I don't see how this behavior can even be happening. There must be some component somewhere in the middle that decides to reset the timestamp to 0.

If you upload over web UI and refresh the page, is the timestamp also 0 ? If yes, open the browser's network console and upload a file again and check the "X-OC-Mtime" request header which should be non-zero.

**Comment by GeneralAmbush** [comment]
Yes, see attached image. But these are no special apps, just the system apps, contacts, calendar etc.
![2017-12-22_owncloud-apps](https://user-images.githubusercontent.com/32911160/34294084-37d7f4ce-e707-11e7-850c-10107a9d4677.png)

When uploading a txt-file over web UI, the "X-OC-Mtime" header seems to be correct, see attachted image. 
![2017-12-22_owncloud_uploading](https://user-images.githubusercontent.com/32911160/34294240-eab079d6-e707-11e7-8e06-5430e32b7e30.png)

Executing `select mtime, storage_mtime from oc_filecache where path like '%test.txt%' `, MySQL returns 3 rows, each row has 'mtime' and 'storage_mtime' zero.

**Comment by PVince81** [comment]
Ok, so this tells us that there is somewhere between the layers that might be setting the value to 0.

Can you check the data folder and see whether the files on disk also have a zero mtime ?
I wonder whether your hoster maybe forbids the usage of the `touch()` system call which is used to set the data.

**Comment by PVince81** [comment]
This patch will add debug logging for the `touch()` operation:
```diff
diff --git a/lib/private/Files/Storage/Local.php b/lib/private/Files/Storage/Local.php
index 44f08c38e9..0fd76eb1a5 100644
--- a/lib/private/Files/Storage/Local.php
+++ b/lib/private/Files/Storage/Local.php
@@ -197,6 +197,7 @@ class Local extends \OC\Files\Storage\Common {
                if ($this->file_exists($path) and !$this->isUpdatable($path)) {
                        return false;
                }
+               \OCP\Util::writeLog('DEBUG', 'Local::touch() $path=' . $path . ' $mtime=' . $mtime, \OCP\Util::DEBUG);
                if (!is_null($mtime)) {
                        $result = touch($this->getSourcePath($path), $mtime);
                } else {
```

Note that you need to set `loglevel => 0` in your config.php.

**Comment by GeneralAmbush** [comment]
Disk files are showing the correct timestamp:
![2017-12-22_ftp](https://user-images.githubusercontent.com/32911160/34302875-30c48bee-e733-11e7-99c9-a07653d9f6cb.png)

Adding your patch and setting loglevel = 0 creates this output:
```json
{
	"reqId": "SrCeg5CHL8JTX9s0Q3sD",
	"level": 0,
	"time": "2017-12-22T15:05:06+00:00",
	"remoteAddr": "remoteAddr",
	"user": "user",
	"app": "DEBUG",
	"method": "PUT",
	"url": "\/owncloud\/remote.php\/webdav\/test.txt",
	"message": "Local::touch() $path=files\/test.txt $mtime=1513955054"
}{
	"reqId": "SrCeg5CHL8JTX9s0Q3sD",
	"level": 3,
	"time": "2017-12-22T15:05:07+00:00",
	"remoteAddr": "remoteAddr",
	"user": "user",
	"app": "PHP",
	"method": "PUT",
	"url": "\/owncloud\/remote.php\/webdav\/test.txt",
	"message": "You are using a fallback implementation of the intl extension. Installing the native one is highly recommended instead. at \/path\/owncloud\/lib\/composer\/patchwork\/utf8\/src\/Patchwork\/Utf8\/Bootup\/intl.php#18"
}{
	"reqId": "PFlKed7DjYU91ODFXWlE",
	"level": 3,
	"time": "2017-12-22T15:05:08+00:00",
	"remoteAddr": "remoteAddr",
	"user": "user",
	"app": "PHP",
	"method": "PROPFIND",
	"url": "\/owncloud\/remote.php\/webdav\/test.txt",
	"message": "You are using a fallback implementation of the intl extension. Installing the native one is highly recommended instead. at \/path\/owncloud\/lib\/composer\/patchwork\/utf8\/src\/Patchwork\/Utf8\/Bootup\/intl.php#18"
}{
	"reqId": "1ke9sTMO1ihpPCzVtyIm",
	"level": 0,
	"time": "2017-12-22T15:05:09+00:00",
	"remoteAddr": "remoteAddr",
	"user": "user",
	"app": "core",
	"method": "GET",
	"url": "\/owncloud\/index.php\/core\/preview.png?file=%2Ftest.txt&c=ad1a7bbc4b727a2e532d733df351b7f9&x=32&y=32&forceIcon=0",
	"message": "Generating preview for \"\/test.txt\" with \"OC\\Preview\\TXT\""
}
```
mtime seems to be correct (1513955054)

**Comment by GeneralAmbush** [comment]
I've done an update to ownCloud 10.0.4 - the error still occurs. mtime is still zero in database.

**Comment by PVince81** [comment]
Ok, so if `touch` works and files on disk have the correct timestamp, maybe the scan method (`stat`) is not receiving the correct mtime when querying it from disk.

Another debug line to check that:
```
diff --git a/lib/private/Files/Storage/Local.php b/lib/private/Files/Storage/Local.php
index 44f08c38e9..6ce1f97705 100644
--- a/lib/private/Files/Storage/Local.php
+++ b/lib/private/Files/Storage/Local.php
@@ -138,6 +138,7 @@ class Local extends \OC\Files\Storage\Common {
                        $statResult['size'] = $filesize;
                        $statResult[7] = $filesize;
                }
+               \OCP\Util::writeLog('DEBUG', 'Local::stat() $path=' . $path . ' $statResult=' . json_encode($statResult), \OCP\Util::DEBUG);
                return $statResult;
        }
```

**Comment by GeneralAmbush** [comment]
Adding the patch and uploading a file test.txt, owncloud.log shows:
```json
{
	"reqId": "BWk8w85mk2CF8ym65lu7",
	"level": 0,
	"time": "2018-01-20T16:14:23+00:00",
	"remoteAddr": "remoteAddr",
	"user": "user",
	"app": "DEBUG",
	"method": "PUT",
	"url": "\/owncloud\/remote.php\/webdav\/test.txt",
	"message": "Local::stat() $path=files\/test.txt $statResult={\"0\":51744,\"1\":7612550,\"2\":33204,\"3\":1,\"4\":3389,\"5\":1582,\"6\":0,\"7\":8296,\"8\":1516464862,\"9\":1516464863,\"10\":1516464863,\"11\":4096,\"12\":24,\"dev\":51744,\"ino\":7612550,\"mode\":33204,\"nlink\":1,\"uid\":3389,\"gid\":1582,\"rdev\":0,\"size\":8296,\"atime\":1516464862,\"mtime\":1516464863,\"ctime\":1516464863,\"blksize\":4096,\"blocks\":24}"
}{
	"reqId": "BWk8w85mk2CF8ym65lu7",
	"level": 0,
	"time": "2018-01-20T16:14:23+00:00",
	"remoteAddr": "remoteAddr",
	"user": "user",
	"app": "DEBUG",
	"method": "PUT",
	"url": "\/owncloud\/remote.php\/webdav\/test.txt",
	"message": "Local::stat() $path=files\/test.txt $statResult={\"0\":51744,\"1\":7612550,\"2\":33204,\"3\":1,\"4\":3389,\"5\":1582,\"6\":0,\"7\":8296,\"8\":1513955054,\"9\":1513955054,\"10\":1516464863,\"11\":4096,\"12\":24,\"dev\":51744,\"ino\":7612550,\"mode\":33204,\"nlink\":1,\"uid\":3389,\"gid\":1582,\"rdev\":0,\"size\":8296,\"atime\":1513955054,\"mtime\":1513955054,\"ctime\":1516464863,\"blksize\":4096,\"blocks\":24}"
}
```

What is the difference between folders and files? Folders seem to be handled different since they have correct mtime. Selecting
```sql
select * from oc_filecache where `mimetype`=2 and `path` like 'files/%' and `mtime` =0
```
has no results.

**Comment by PVince81** [comment]
@GeneralAmbush folder mtime is not taken from the file I guess, it likely just takes the current system time.

The debug output above looks correct, the `stat()` method returns the correct mtime value.
This is really strange, should not even be possible.

**Comment by GeneralAmbush** [comment]
@PVince81 Seems strange for me, too. Any further ideas?

**Comment by PVince81** [comment]
@GeneralAmbush I've run out of ideas and time,sorry. The fact that you're the only person who reported this so far seem to point at the possibility that this is an environment issue of some sorts.

Maybe you could switch to a different shared hoster or ask for support on central: http://central.owncloud.org/

**Comment by fhruschka** [comment]
@PVince81: GeneralAmbush is not the only case.
I have exactly the same problem with version 10.0.4, while 9.1.6 works fine. (I've both versions running on the same server). I'm using Sync Client 2.4.0

Uploaded files have the correct date in the data folder (checked with FTP) but show modified "seconds ago" in the web interface of 10.0.4 and sync to other devices with date 01/01/1970.

**Comment by PVince81** [comment]
Do you both appear to be on the same shared hosting provider ?

**Comment by fhruschka** [comment]
In my case, I have a dedicated managed Debian server at Internex, an Austrian IT company.

**Comment by PVince81** [comment]
Do you guys have access to the php.ini settings ? Maybe these will contain more clues...

**Comment by fhruschka** [comment]
Please find my php settings here:
#https://#owncloud.hruschka.com/info.#php
(delete the hashes to make it a working URL)

**Comment by PVince81** [comment]
disabled functions: `escapeshellcmd, eval, exec, passthru, popen, proc_open, shell_exec, show_source, system`

hmm that shouldn't be a problem

**Comment by fhruschka** [comment]
Disabled unless really needed.

**Comment by PVince81** [comment]
@fhruschka are you able to temporarily allow all these, just in case ? At least it would iron out the possibility that blocking these is not the cause of this issue.

**Comment by fhruschka** [comment]
Hmm, only by a support request to my provider. 

As said, it is a managed server, and for very good reasons IT staff from Internex keeps amateurs like me away from configuration. ;-)  It is a production server any my business depends on its reliability.

Of course, I can ask them for any feature, but they always check for security of the system if I really need it.
I'd prefer to avoid "experiments", sorry.

**Comment by PVince81** [comment]
No worries, I understand.

To be able to find out what happens it would required attaching a debugger to the PHP code and find out what is actually happening. The log entries above already show that everything should be correct and yet the values look wrong.

**Comment by fhruschka** [comment]
That would be no problem. 
I can do changes to the OC10 php scripts, and if it is of use, I can also add you as a user to the cloud.

**Comment by GeneralAmbush** [comment]
@PVince81: I'm hosting owncloud on a managed server at tiggersWelt in germany (http://tiggerswelt.net).

**Comment by GeneralAmbush** [comment]
Hey guys,
I made contact to my webhoster. He doesn't have a clue why it is not working on the environment.

function filemtime($path) in OC\Files\Storage\Local:
```php
public function filemtime($path) {
		$fullPath = $this->getSourcePath($path);
		clearstatcache($fullPath);
		if (!$this->file_exists($path)) {
			return false;
		}
		if (PHP_INT_SIZE === 4) {
			if (\OC_Util::runningOn('linux')) {
				return (int) exec ('stat -c %Y '. escapeshellarg ($fullPath));
			} else if (\OC_Util::runningOn('bsd') || \OC_Util::runningOn('mac')) {
				return (int) exec ('stat -f %m '. escapeshellarg ($fullPath));
			}
			return false;
		}
		return filemtime($fullPath);
	}
```
Analysing the function points out the following:
- With 'PHP_INT_SIZE === 4', the function is checking for a 32-bit integer php (which is used in my environment)
- in this case, the function falls back to exec ('stat ...'). @fhruschka: this might be the point where the error occurs in your environment because of the forbidden exec (if you are using a 32-bit-php)
- in my case, exec is not forbidden, but there is no 'stat' in the shell php is using (busybox)

The point is: Why is owncloud using this 'workaround for 32-bit-php'? Why not using filemtime() directly? In fact, the result of 'stat()' is casted to (int) in 32-bit.

Removing the whole workaround and just using filemtime() and uploading a file via web-ui, the mtime in database is now !=0. In the ui and in windows explorer (synced via windows client), the timestamp is no longer 01/01/1970. 
The strange thing is that the mtime now is incorrect: Uploading a few minutes ago, it's written as 1517516854 (22.12.2017 - 16:04:14).

Nevertheless, I hope this helps to find a solution!

**Comment by fhruschka** [comment]
@GeneralAmbush Thanks for this. The same function in Version 9 (which works fine) does not use the exec function; and there are apparently (according to a text search) only 2 php scripts that use it; one of them is is the file Local.php which you mention.

This is a good rationale to ask my provider to enable it. (better than "please enable everything" :-)
I'll do so and report back

**Comment by PVince81** [comment]
It is possible that older PHP versions or PHP on some setups with 32-bit (not all of them) do not return correct values for `filemtime`.

Maybe the workaround should first check if `exec` is available... or we should remove it completely...

Might need a bit of digging in git history to find the exact reason.

**Comment by PVince81** [comment]
Seems this `exec` thing was added here: https://github.com/owncloud/core/pull/25428.

Are you guys all running on 32-bit systems ?

**Comment by GeneralAmbush** [comment]
Yes, I'm running owncloud on an 32-bit system.

@PVince81: You suggested to check whether `exec` is available. I would prefer removing the 32-bit-workaround completely if there is no valid reason to keep it. In my case, `exec` is available, but I have no `stat` so `exec ('stat -c %Y '. escapeshellarg ($fullPath));` returns no result.

I'll check the fact that `filemtime` is returning an incorrect value with my webhoster.

**Comment by fhruschka** [comment]
My provider internex.at did the configuration immediately, but I was out of office this morning. Apologies therefore for the delayed feedback:

The General's hint was right. **With exec enabled, filetimes are properly set**.
In my setup it seems to work fine for files uploaded through the web UI, as well as through the sync client.

I would suggest to go back to the way this was done in version 9 and earlier, which apparently did not use the exec function and worked fine. I use owncloud since version 1.1.3 and never had such problems. I also have this managed server at Internex now for almost 10 years, and none of the web applications I used during all these years ever needed the exec function.

**Comment by PVince81** [comment]
> worked fine

It didn't work fine for the person who submit the PR. Not sure what was so special about their env. (I didn't investigate to find out)

Best would be to have a solution that works for both. Maybe add a check after `exec` and fall back to `filemtime`

**Comment by fhruschka** [comment]
ok, right, apologies!
Thanks

**Comment by PVince81** [comment]
Can you guys try uploading a 4 GB file ? with the `filemtime` method ? It is likely that this will fail and is likely the reason the `exec+stat` method was added in the first place.

**Comment by fhruschka** [comment]
will do so

**Comment by fhruschka** [comment]
@PVince81: You were right, it failed. 

I'm not completely sure what you meant with "_with the filemtime method_"; I just let the sync client upload a 4GB file and when it had finished uploading it finally aborted with the message 
> Chunks on server do not sum up to 4238222336 but to 4238222336

Hope this helps

**Comment by PVince81** [comment]
@fhruschka did you modify your code or does it still have the `exec()` in place ?

With the `filemtime` method I mean, look at the code from https://github.com/owncloud/core/issues/29287#issuecomment-362392881 and comment out the whole block of `if (PHP_INT_SIZE == 4) { ... }` so the check will use the `filemtime()` call below. Then upload a 4 GB file.

Your current result about chunks not summing up looks like this other issue: https://github.com/owncloud/core/issues/29447

**Comment by fhruschka** [comment]
@PVince81 That was with the original 10.0.4 version, having the `exec` function

But I think to remember that during installation the script gave some warning about +4GB files ... which I ignored because I never use OC for files of such size. I therefore really don't remember what exactly it was. Maybe it re-appears when I update to 10.0.6. So this is probably a known issue.

**Comment by bitbaendiger** [comment]
Hello there,

I'm responsible for the hosting-environment of @GeneralAmbush and did some further investigation on this topic. At the moment I don't see any problem with `filemtime()` on 32-Bit Environments - at least on ours. As `filemtime()` returns a timestamp I would guess it's affacted by a year 2038 problem, but not any size-related. Unfortunately I do not have an ARM-Plattform available to reconstruct the problems mentioned in #25428.

To gather some environment-specific data I've created a file sized 4GB + 1MB file using dd and had PHP read some values

On 32-Bit:
~~~ {.bash}
# dd if=/system/dev/zero of=4GB bs=1024 count=$((4096*1024+1024))
# php -a
Interactive shell

php > echo PHP_INT_SIZE;
4
php > echo filesize ('4GB');
1048576
php > echo filemtime ('4GB');
1518175076
php > var_dump (stat ('4GB'));
array(26) {
  [...]
  ["size"]=>
  int(1048576)
  ["atime"]=>
  int(1518174999)
  ["mtime"]=>
  int(1518175076)
  ["ctime"]=>
  int(1518175076)
  ["blksize"]=>
  int(4096)
  ["blocks"]=>
  int(8390664)
}
~~~

On 64-Bit (not released to production-use here yet):
~~~ {.bash}
# dd if=/system/dev/zero of=4GB bs=1024 count=$((4*1024*1024+1024))
# php -a
Interactive shell

php > echo PHP_INT_SIZE;
8
php > echo filesize ('4GB');
4296015872
php > echo filemtime ('4GB');
1518176071
php > var_dump (stat ('4GB'));
array(26) {
  [...]
  ["size"]=>
  int(4296015872)
  ["atime"]=>
  int(1518176063)
  ["mtime"]=>
  int(1518176071)
  ["ctime"]=>
  int(1518176071)
  ["blksize"]=>
  int(4096)
  ["blocks"]=>
  int(8390664)
}
~~~

Timestamps are different because of different systems and because tests were not carried out in parallel. The only broken value is "size" from `stat()` (and from `filesize()` which is the same source), which may be detected by checking with `((($stat ['blocks'] - 1) * 512) < $stat ['size'])` (using float, not int, which is implied by PHP).

As I'm not familiar with the ARM-Issue mentioned above I would suggest checking `$return_var` of `exec()` before returning it's output as int, `if ($return_var != 0)` OC could fallback to `filemtime()` which should work in both reported cases here well. `stat` is not part of LSB therefore I would not rely on it to be present.

**Comment by PVince81** [comment]
Thanks for the research. @sharidas do you mind taking care of implementing the `exec` check ?

**Comment by PVince81** [comment]
can you guys try with https://github.com/owncloud/core/pull/30546 ?

**Comment by GeneralAmbush** [comment]
hey guys,
thanks for taking care. 

I've created some test-files and uploaded them to my OC with #30546 applied. 

UI shows the correct timestamp ("a few month ago" and "4 minutes ago") and desktop client is syncing them correctly. mtime and storage_mtime seem to be correct in mysql.

conclusion: works for me!

**Comment by PVince81** [comment]
Great, thanks for the feedback!

The fix will be in the upcoming 10.0.8 version.

I'm closing this then.
